home/farcall.asm

```diff
@@ -2,131 +2,233 @@
-; Updates the flags on the stack to their current value
-; without clobbering any other registers.
-	push af ; pushes current flags
+_RstFarCall::
+	call .do_farcall
+	jr _ReturnFarCall
+; Call the following dab pointer
+; preserves af, bc, de, hl
+
+; Stack layout
+; +11 return address
+; +10 saved bank
+; +8  return path
+; +6  function pointer
+; +4  saved af
+; +2  saved de
+; +0  saved hl
+.do_farcall
+; same speed and size as add sp, -2 but preserves flags
+	dec sp
+	dec sp
+	push af
+	push de
+	push hl
+
+	ld hl, sp + 10
+	ldh a, [hROMBank]
+	ld [hli], a
+
+; Read return address
+	ld a, [hli]
+	ld h, [hl]
+	ld l, a
+
+; function pointer
+	ld a, [hli]
+	ld e, a
+	ld a, [hli]
+	ld d, a
+
+; check farjp
+	add a
+
+; bank number
+	ld a, [hli]
+	rst Bankswitch
+
+	jr nc, .call
+	res 7, d
+; for farjp we need to replace the return address
+	ld hl, DoNothing
+.call
 	push hl
-	ld hl, sp + 2
-	ld a, [hli] ; reads current flags
-	assert HIGH(wStackBottom) == HIGH(wStackTop)
-	inc l
-	ld [hl], a ; overwrites pushed flags
+
+; write the function pointer
+	ld hl, sp + 8
+	ld a, e
+	ld [hli], a
+	ld [hl], d
+
+	pop de
+; write return address
+	ld hl, sp + 11
+	ld a, e
+	ld [hli], a
+	ld [hl], d
+
+; restore registers and jump to the target function
 	pop hl
+	pop de
 	pop af
-ENDM
-
-RstFarCall::
-; Call the following dba pointer.
-	ldh [hFarCallSavedA], a
-	ld a, h
-	ldh [hFarCallSavedH], a
-	ld a, l
-	ldh [hFarCallSavedL], a
-	pop hl
+	ret
+
+FarPointerCall::
+; Call the dba pointer at hl. Clobbers a and hl.
 	ld a, [hli]
-	ldh [hTempBank], a
+	push af
 	ld a, [hli]
-	bit 7, [hl]
-	jr nz, .farjp
-	inc hl
-	push hl
-	dec hl
-.farjp
 	ld h, [hl]
 	ld l, a
-	res 7, h
+	pop af
+; fallthrough
+
+FarCall_hl::
+; Call a:hl.
+; clear carry
+	and a
 _DoFarCall:
-	ldh a, [hROMBank]
+	dec sp
+	push hl
 	push af
-	ldh a, [hTempBank]
-_DoFarCall_BankInA:
-	rst Bankswitch
-	call RetrieveAHLAndCallFunction
-_ReturnFarCall:
-	ldh [hFarCallSavedA], a
-	update_pushed_af_flags
+	ldh a, [hROMBank]
+	ld hl, sp + 4
+	ld [hl], a
 	pop af
 	rst Bankswitch
-	ldh a, [hFarCallSavedA]
-	ret
-
-RunFunctionInWRA6::
-; Call the following function in wDecompressScratch's WRAM bank. Clobbers a.
-	ld a, BANK(wDecompressScratch)
-StackCallInWRAMBankA::
-; Call the following function in WRAM bank a. Clobbers a.
-	ldh [hTempBank], a
-	ld a, h
-	ldh [hFarCallSavedH], a
-	ld a, l
-	ldh [hFarCallSavedL], a
 	pop hl
-	ldh a, [rSVBK]
+	jr nc, .hl
+	call _de_
+	jr _ReturnFarCall
+.hl
+	call _hl_
+; fallthrough
+_ReturnFarCall:
+; restore the rom bank and clean up
 	push af
-	ldh a, [hTempBank]
-	ldh [rSVBK], a
-	call RetrieveAHLAndCallFunction
-	ldh [hTempBank], a
-	update_pushed_af_flags
+	push hl
+	ld hl, sp + 4
+	ld a, [hl]
+	rst Bankswitch
+	pop hl
 	pop af
-	ldh [rSVBK], a
-	ldh a, [hTempBank]
+	inc sp
 	ret
 
-RetrieveAHLAndCallFunction:
-; Call the function at hl with restored values of a and hl.
+FarCall_de::
+; Call a:de.
+	scf
+	jr _DoFarCall
+
+; Stack layout:
```

Comment:
It's not clear at what point this layout is supposed to actually be the case during all the inc/dec sps and push/pops below.